### ISSUE TITLE
Training time

### ISSUE THREAD

**Issue by dreamflasher** [LeeDoYup/FixMatch-pytorch#1]
The training time of 16 hours for CIFAR10 sounds like a lot – is this for all experiments summed, or a single training (with how many labeled data points)?

Whats the bottleneck? (Supervised CIFAR10 training is in the magnitude of minutes)

**Comment by LeeDoYup** [comment]
Thanks for making the FIRST issue !
Yes, the training time of WRN-28-2 for CIFAR10 takes 16 hours.
The training time is not different by the amount of labeled data, because all training samples are used for training regardless of labeling.
In addition, we check there is no computational bottleneck in training.
We also check that other implementations require similar time to training, 
and our code is more faster than [another code](https://github.com/kekmodel/FixMatch-pytorch), which supports DDP in pytorch.

I agree that consistency regularization based SSL methods take much long time to train the model.
The fundamental reason is that FixMatch does not use external dataset and a pretraining model, but makes hidden representations from the unlabeled data of downstream task (CIFAR10).
In addition, FixMatch requires 2^20 iterations, which are much much longer than those of supervised learning (150 epochs with batch_size = 128, it is about 60,000 iterations).

When the amount of labels is enough in the easy task (e.g. 4000 labels of CIFAR10),
plausible accuracy is reached at the early stage of training (about at < 100,000 iterations?).
However, to improve the accuracy, FixMatch requires remaining iterations (about 900,000).

**Comment by dreamflasher** [comment]
Thank you kindly for the detailed explanation!

And thank you very much for your great work, highly appreciated a clean and well structured pytorch implementation of FixMatch!

**Comment by LeeDoYup** [comment]
We are on testing this branch `dev/speedup` https://github.com/LeeDoYup/FixMatch-pytorch/tree/dev/speedup .
We expect that it increases the speed of training about 15 % w/o performance changes.
After checking all the experiments, we will merge it to the master branch.
(Meanwhile, we already know the computation is not logically different in pytorch, and if you need it, feel free to use the branch.)
